apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/model/ArchiveApiTypeContainer.java

```diff
@@ -149,31 +122,54 @@ public String toString() {
 	String fLocation;
 
 	/**
-	 * Cache of package names to class file paths in that package, or
-	 * <code>null</code> if not yet initialized.
+	 * Cache of package names to a map of class names to class files paths in that
+	 * package, or <code>null</code> if not yet initialized.
 	 */
-	private Map<String, Set<String>> fPackages;
+	private Map<String, Map<String, String>> fPackages;
 
 	/**
 	 * Cache of package names in this archive.
 	 */
 	private String[] fPackageNames;
 
 	/**
-	 * Constructs an {@link IApiTypeContainer} container for the given jar or
-	 * zip file at the specified location.
+	 * Constructs an {@link IApiTypeContainer} container for the given jar or zip
+	 * file at the specified location.
 	 *
 	 * @param parent the parent {@link IApiElement} or <code>null</code> if none
-	 * @param path location of the file in the local file system
+	 * @param path   location of the file in the local file system
 	 */
 	public ArchiveApiTypeContainer(IApiElement parent, String path) {
 		super(parent, IApiElement.API_TYPE_CONTAINER, path);
 		this.fLocation = path;
-		if (path.endsWith("jrt-fs.jar")) { //$NON-NLS-1$
-			IPath newPath = new Path(path);
-			newPath = newPath.removeLastSegments(2).addTrailingSeparator();
-			newPath = newPath.append("jmods").append("java.base.jmod"); //$NON-NLS-1$ //$NON-NLS-2$
-			this.fLocation = newPath.toOSString();
+	}
+
+	/**
+	 * Converts the location to a path in the applicable file system.
+	 *
+	 * @return the path corresponding to the location.
+	 * @throws IOException
+	 */
+	@SuppressWarnings("nls")
+	private Path getLocation() throws IOException {
+		if (fLocation.endsWith("jrt-fs.jar")) {
+			AtomicReference<IOException> exception = new AtomicReference<>();
+			FileSystem jrtFileSystem = JRTS.computeIfAbsent(fLocation, it -> {
+				Path jrePath = Path.of(fLocation).getParent().getParent();
+				try {
+					return FileSystems.newFileSystem(URI.create("jrt:/"),
+							Collections.singletonMap("java.home", jrePath.toString()));
+				} catch (IOException e) {
+					exception.set(e);
```

Comment:
If the file system already exists, you can get it in the catch block.

Reply:
Given that it is quite hard to get it right with all the different cases can't PDE reuse the code from JDT here?

Reply:
> If the file system already exists, you can get it in the catch block.

Each call creates a new instance.   Really, I've debugged into this, including looking at how many instances of jdt.internal.jrtfs.JrtFileSystem are created (and there are a significant number, but from other locations).  

It did also did occur to me to look at how JDT does this, which I mentioned here:

https://github.com/eclipse-pde/eclipse.pde/pull/357#discussion_r992300456

You can look at the code here:

https://github.com/eclipse-jdt/eclipse.jdt.core/blob/9562c376de84b672f4925bf2f49c10b842073f0f/org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/util/JRTUtil.java#L502-L512

I did also test locally with a JRE and and JDK and the tests did all pass.

Even JDT seems to do this in multiple places, e.g., also in org.eclipse.jdt.internal.compiler.tool.JrtFileSystem.initialize().  But none of it very reusable because no one wants to over-commit on APIs...

So I believe there are no outstanding issues. 

Certainly if someone likes they may change JDT to provide something reusable and then make downstream changes at a later point in time.

Reply:
The point is not that **this** code can create new file system, but **other** code (JDT) too. So the pattern is to "try create new / catch already there? just get it"

Reply:
No, please look at the code.
```
	JrtFileSystem(Jdk jdkHome, String release) throws IOException {
		this.jdk = jdkHome;
		this.release = release;
		JRTUtil.MODULE_TO_LOAD = System.getProperty("modules.to.load"); //$NON-NLS-1$
		HashMap<String, String> env = new HashMap<>();
		env.put("java.home", this.jdk.path); //$NON-NLS-1$
		this.fs = FileSystems.newFileSystem(JRTUtil.JRT_URI, env);
		this.modRoot = this.fs.getPath(JRTUtil.MODULES_SUBDIR);
		// Set up the root directory where modules are located
		walkJrtForModules();
	}
```

Or at this org.eclipse.jdt.internal.compiler.tool.JrtFileSystem.initialize() code:
```
	public void initialize() throws IOException {
		// initialize packages
		this.modulePathMap = new HashMap<>();
		URL jrtPath = null;

		if (this.file.exists()) {
			jrtPath = Paths.get(this.file.toPath().toString(), "lib", JRTUtil.JRT_FS_JAR).toUri().toURL(); //$NON-NLS-1$
			try (URLClassLoader loader = new URLClassLoader(new URL[] { jrtPath })) {
				HashMap<String, ?> env = new HashMap<>();
				this.jrtfs = FileSystems.newFileSystem(JRT_URI, env, loader);
				this.modules = this.jrtfs.getPath("/modules"); //$NON-NLS-1$
			}
		} else {
			return;
		}

		org.eclipse.jdt.internal.compiler.util.JRTUtil.walkModuleImage(this.file,
				new org.eclipse.jdt.internal.compiler.util.JRTUtil.JrtFileVisitor<Path>() {
			@Override
			public FileVisitResult visitModule(Path path, String name) throws IOException {
				JrtFileSystem.this.modulePathMap.put(name, path);
				return FileVisitResult.CONTINUE;
			}
		}, JRTUtil.NOTIFY_MODULES);
	}
```
I suspect you are confusing those with this pattern used for the jar/zip file system:
```
	public static FileSystem getJarFileSystem(Path path) throws IOException {
		URI uri = URI.create("jar:file:" + path.toUri().getRawPath()); //$NON-NLS-1$
		try {
			try {
				return FileSystems.getFileSystem(uri);
			} catch (FileSystemNotFoundException fne) {
				try {
					return FileSystems.newFileSystem(uri, Map.of(), ClassLoader.getSystemClassLoader());
				} catch (FileSystemAlreadyExistsException e) {
					return FileSystems.getFileSystem(uri);
				}
			}
		} catch (ProviderNotFoundException e) {
			throw new IOException("No provider for uri " + uri, e); //$NON-NLS-1$
		}
	}
```
But that's not what JDT does for the JRT file system in the **three** places (also org.eclipse.jdt.internal.compiler.apt.util.JrtFileSystem.initialize()) that it does this. Note too that unlike the zip/jar file system where there are unique URIs involved, all these places the URI  'jrt:/' and create different instances depending on the java.home value in the environment.  So they cannot use the FileSystems.getFileSystem() pattern...

So again, all these create their own new instance, none of that share that instance, and none of them catch the exception when creating it.  So I believe this is not an issue....

Reply:
Well, there is lot of various uses of the API, and not all of them are correct or intended for general use in the IDE. JDT code in compiler.tool is from command line ecj compiler, which wasn't meant to be used in the multi-threaded environment of an IDE, where every one can create file system objects with any (also same) paths.

PDE code however **is** in the IDE, so it must be aware that there can be someone already created the same file system or creating it in parallel at same time.

I would recommend to look at `org.eclipse.jdt.internal.compiler.util.CtSym.init()`

```
try {
			fst = FileSystems.getFileSystem(uri);
		} catch (Exception fne) {
			// Ignore and move on
		}
		if (fst == null) {
			try {
				fst = FileSystems.newFileSystem(uri, new HashMap<>(), ClassLoader.getSystemClassLoader());
			} catch (FileSystemAlreadyExistsException e) {
				fst = FileSystems.getFileSystem(uri);
			} catch (ProviderNotFoundException e) {
				throw new IOException("Failed to create ct.sym file system for " + this.ctSymFile, e); //$NON-NLS-1$
			}
		}
```

Reply:
Also please note the classloader given as third argument. It's important too.

Reply:
I'm not sure why you are pointing me to another another "jar:" example?  I already mentioned that **this is a different pattern** because each jar has a different URI, but the JRT file system **has only one** URI "jre:/"

If the class loader is important,  then surely org.eclipse.jdt.internal.compiler.util.JrtFileSystem.JrtFileSystem(Jdk, String) should use it to, but it doesn't.

The class loader is used only to load the provider in java.nio.file.FileSystems.newFileSystem(URI, Map<String, ?>, ClassLoader):
```
    public static FileSystem newFileSystem(URI uri, Map<String,?> env, ClassLoader loader)
        throws IOException
    {
        String scheme = uri.getScheme();

        // check installed providers
        for (FileSystemProvider provider : FileSystemProvider.installedProviders()) {
            if (scheme.equalsIgnoreCase(provider.getScheme())) {
                try {
                    return provider.newFileSystem(uri, env);
                } catch (UnsupportedOperationException uoe) {
                }
            }
        }

        // if not found, use service-provider loading facility
        if (loader != null) {
            ServiceLoader<FileSystemProvider> sl = ServiceLoader
                .load(FileSystemProvider.class, loader);
            for (FileSystemProvider provider : sl) {
                if (scheme.equalsIgnoreCase(provider.getScheme())) {
                    try {
                        return provider.newFileSystem(uri, env);
                    } catch (UnsupportedOperationException uoe) {
                    }
                }
            }
        }

        throw new ProviderNotFoundException("Provider \"" + scheme + "\" not found");
    }
```
Do we really want to load classes from the target JRE/JDK?  I'm a little doubtful of that...

Reply:
@iloveeclipse I assume that silence is implicit agreement...

Reply:
> @iloveeclipse I assume that silence is implicit agreement...

I assume you have tested that & I assume the "jrt:/" must be a special case, the typical behavior I'm used to see is that JVM prevents multiple instances of file systems mapped to same url to be created. So feel free to forget my comments above :-)